src/lib.rs

```diff
@@ -0,0 +1,459 @@
+
+        // Register validation function
+        fn new_registry(origin, validation_fn_addr: T::AccountId) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+
+            // Reserve fee for token
+            <T as Trait>::Currency::reserve(&sender, T::NFTValidationRegistryDeposit::get())?;
+
+            // Keep old value for event
+            let uid = NextRegistryId::get();
+
+            // Write state
+            <ValidationFn<T>>::insert(&uid, validation_fn_addr);
+
+            // Update next registry id
+            NextRegistryId::mutate(|value| *value += 1);
+
+            // Events
+            Self::deposit_event(RawEvent::NewRegistry(sender, uid));
+
+            Ok(())
+        }
+
+        // Mint a new token
+        // Value: account can transfer some currency to smart contract via calling
+        // Gas limit: set the maximum gas usage for smart contract execution in WASM
+        fn mint(origin,
+            uid: RegistryUid,
+            token_id: T::Hash,
+            metadata: Vec<u8>,
+            anchor_id: T::Hash,
+            pfs: Vec<Proof>,
+            static_proofs: [H256;3],
+            value: contracts::BalanceOf<T>,
+            gas_limit: contracts::Gas,
+        ) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+
+            // Contract registered for the uid
+            Self::ensure_validation_fn_exists(uid)?;
+
+            // Ensure token id not existed
+            Self::ensure_token_not_existed(&token_id)?;
+
+            // Ensure metadata is valid
+            Self::ensure_metadata_valid(&metadata)?;
+
+            // Get the doc root
+            let doc_root = Self::get_document_root(&anchor_id)?;
+
+            // Verify the proof against document root
+            Self::validate_proofs(&doc_root, &pfs, &static_proofs)?;
+
+            // Get the hash of validate method in contract
+            let keccak = ink_utils::hash::keccak256("validate".as_bytes());
+
+            // Encode 4 bytes of the method name's hash
+            let selector = [keccak[0], keccak[1], keccak[2], keccak[3]];
+            let mut call = selector.encode();
+
+            // Compute the bundled hash
+            let proof_leaves = pfs.iter().map(|proof| proof.leaf_hash).collect();
+
+            // Put parameters into single struct.
+            let contract_parameter = ContractParameter::<T::Hash, T::AccountId> {
+                uid: uid,
+                token_id: token_id,
+                token_owner: sender.clone(),
+                metadata: metadata.clone(),
+                proof_leaves: proof_leaves,
+            };
+
+            // Append the parameter after method
+            call.append(&mut Encode::encode(&contract_parameter));
+
+            // Call the contract via bare call
+            <contracts::Module<T>>::bare_call(
+                sender,
+                Self::validator_of(uid).unwrap(),
```

Comment:
I will restructure the code to avoid the unwrap call. But to my knowledge, the extrinsics will be executed in serials like Ethereum. I am not sure about if substrate can support execute in parallel. I just heard it may be supported in the far future.

Reply:
Thanks! It's a small change to prevent any potential panics and it prevents big headaches in the future, so it makes sense to change them now.